Totp enrollment frontend

CDimonaco opened the pull request:
# Description

This pr implements the totp enrollment frontend procedure.

![Recording 2024-05-30 at 10 50 35](https://github.com/trento-project/web/assets/9409502/4ee50f43-2f2b-41e1-ab90-01e39e62dab1)


## How was this tested?

Automated tests

abravosuse reviewed (approved):
Based on the clip in the description and the storybook, it looks good to me.

arbulu89 reviewed (commented):
Hey @CDimonaco ,
The code looks good.
I think that we should have a confirmation modal to disable the TOTP.
Otherwise just clicking the switch disables it, and it looks to vulnerable to that.
Because I wouldn't expect just to disable it without any confirmation.
What do you think?

The rest of the changes are done when you click the `Save` button, so having this different would create confusion.

CDimonaco commented:
@arbulu89 You are completely right, I forgot the confirmation modal, back to draft 🚀

EMaksy reviewed (commented):
Hey man so far so good, I just left some small comments, going to review after you add the confirmation modal

CDimonaco commented on assets/js/lib/test-utils/factories/users.js:
```diff
@@ -30,6 +30,7 @@ export const profileFactory = Factory.define(() => ({
   password_change_requested: false,
   created_at: formatISO(faker.date.past()),
   updated_at: formatISO(faker.date.past()),
+  totp_enabled: faker.datatype.boolean(),
```
done

CDimonaco commented on assets/js/pages/Profile/TotpEnrollmentBox.jsx:
```diff
@@ -0,0 +1,67 @@
+          <Input
+            autoComplete="off"
+            placeholder="TOTP code"
+            name="totp_code"
```
The api field is called totp_code, I will stay with the totp code so the form semantic of setting the field is mantained

CDimonaco commented on assets/js/pages/Profile/TotpEnrollmentBox.stories.jsx:
```diff
@@ -0,0 +1,56 @@
+        type: 'text',
+      },
+    },
+    verifyError: {
```
added nice catch

CDimonaco commented on assets/js/pages/Profile/TotpEnrollmentBox.test.jsx:
```diff
@@ -0,0 +1,81 @@
+    expect(screen.getByText('Required field')).toBeVisible();
+  });
+
+  it('should errors to the field when provided', async () => {
```
done

CDimonaco commented on assets/js/pages/Profile/ProfileForm.jsx:
```diff
@@ -5,21 +5,31 @@ import Button from '@common/Button';
   onSave = noop,
+  onResetTotp = noop,
+  onVerifyTotp = noop,
+  onTotpEnable = noop,
```
done

CDimonaco commented on assets/js/pages/Profile/ProfilePage.jsx:
```diff
@@ -28,13 +38,70 @@ function ProfilePage() {
     setErrors([]);
   };
 
+  const verifyTotpEnrolling = async (enrollmentTotp) => {
```
done

CDimonaco commented on assets/js/pages/Profile/ProfilePage.jsx:
```diff
@@ -28,13 +38,70 @@ function ProfilePage() {
+      setSaving(false);
+      toast.success('TOTP Disabled');
+    } catch {
+      toast.error('Error disabling totp, please refresh profile.');
```
done

CDimonaco commented:
Confirmation modal added, addressed all the feedbacks

@EMaksy I can agree on the factories comment, but I need also to change the previous test, so I will leave for another pr!

arbulu89 commented on assets/js/pages/Profile/ProfileForm.test.jsx:
```diff
@@ -110,4 +110,192 @@ describe('ProfileForm', () => {
   });
+
+  it('should set the authenticator app switch when totpEnabled is true', async () => {
+    const username = faker.internet.userName();
```
Yes, I do agree. Let's better use the `profileFactory` as it will return you all these fields

arbulu89 commented on assets/js/pages/Profile/ProfilePage.jsx:
```diff
@@ -28,13 +38,70 @@ function ProfilePage() {
+        },
+      },
+    }) {
+      setErrors([{ ...error, source: { pointer: '/totp_code' } }]);
```
I don't see the comment, but nevermind hehe

arbulu89 reviewed (approved):
Well done!
I just miss maybe updating the `ProfileForm.stories.jsx` to add the new fields.
Otherwise storybook sets them as objects, making the usage a bit clunky. Either way, the disable modal for example works in the story hehe
If you do that it would be perfect!

CDimonaco commented:
@arbulu89 Thx, commented added and storybook updated!
